clientconn_state_transition_test.go

```diff
@@ -0,0 +1,428 @@
+	}
+	defer lis.Close()
+
+	sawReady := make(chan struct{})
+
+	// Launch the server.
+	go func() {
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings frame. %v", err)
+			return
+		}
+
+		// Prevents race between onPrefaceReceipt and onClose.
+		<-sawReady
+
+		conn.Close()
+	}()
+
+	client, err := DialContext(ctx, lis.Addr().String(), WithWaitForHandshake(), WithInsecure(), WithBalancerName(stateRecordingBalancerName))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	timeout := time.After(5 * time.Second)
+
+	for i := 0; i < len(want); i++ {
+		select {
+		case <-timeout:
+			t.Fatalf("timed out waiting for state %d (%v) in flow %v", i, want[i], want)
+		case seen := <-stateNotifications:
+			if seen == connectivity.Ready {
+				close(sawReady)
+			}
+			if seen != want[i] {
+				t.Fatalf("expected to see %v at position %d in flow %v, got %v", want[i], i, want, seen)
+			}
+		}
+	}
+}
+
+// When the first connection is closed, the client enters stays in CONNECTING until it tries the second
+// address (which succeeds, and then it enters READY).
+func TestStateTransitions_TriesAllAddrsBeforeTransientFailure(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	want := []connectivity.State{
+		connectivity.Connecting,
+		connectivity.Ready,
+	}
+
+	stateNotifications := make(chan connectivity.State, len(want))
+	testBalancer.ResetNotifier(stateNotifications)
+	defer close(stateNotifications)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	lis1, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis1.Close()
+
+	lis2, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis2.Close()
+
+	server1Done := make(chan struct{})
+	server2Done := make(chan struct{})
+
+	// Launch server 1.
+	go func() {
+		conn, err := lis1.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+
+		conn.Close()
+		close(server1Done)
+	}()
+	// Launch server 2.
+	go func() {
+		conn, err := lis2.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings frame. %v", err)
+			return
+		}
+		close(server2Done)
+	}()
+
+	rb := manual.NewBuilderWithScheme("whatever")
+	rb.InitialAddrs([]resolver.Address{
+		{Addr: lis1.Addr().String()},
+		{Addr: lis2.Addr().String()},
+	})
+	client, err := DialContext(ctx, "this-gets-overwritten", WithInsecure(), WithWaitForHandshake(), WithBalancerName(stateRecordingBalancerName), withResolverBuilder(rb))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	timeout := time.After(2 * time.Second)
```

Comment:
Why not 5s like the others?  2s does seem a bit short in case Travis is having a bad day.  (Another 2s is below.)

Reply:
Done